### PR TITLE
[MAJOR] Do not overwrite all targets with same id (minimal)

### LINKED ISSUE #538: Cordova overwrites bundleIdentifier for all existing targets
# Bug Report
When cordova (using 5.0.0) projects contains watch targets (watchkit, watch extension) which have their own bundle id setup correctly, cordova overwrites all targets with the same id.

## Problem
Its not possible to pack new targets in the project and use cordova cli build 
cordova build --platform ios --device --release

### What is expected to happen?
It should not overwrite the PRODUCT_BUNDLE_IDENTIFIER of all targets.

### What does actually happen?
Signing does not work anymore.

- [ x] I searched for existing GitHub issues
- [ x] I updated all Cordova tooling to most recent version
- [x ] I included all the necessary information above

---

Updates by @brodybits:

This issue seems to also break functionality of plugins that do extensions to Xcode projects by using hooks. See below for discussions.

### PR DESCRIPTION
**Pull request by NiklasMerz**
### Platforms affected
iOS

### Motivation and Context

This is a simplified version of #545 with the minimal changes to fix #538 

### Description

See discussion in #545 

### Testing

Manual tests with complex project

### Checklist

- [ ] I've run the tests to see all new and existing tests pass
- [ ] I added automated test coverage as appropriate for this change
- [ ] Commit is prefixed with `(platform)` if this change only applies to one platform (e.g. `(android)`)
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above (and used the correct [keyword to close issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/))
- [ ] I've updated the documentation if necessary

### CODE REVIEW

File: bin/templates/scripts/cordova/lib/prepare.js [hunk 1]
```diff
@@ -274,12 +274,54 @@ function handleOrientationSettings (platformConfig, infoPlist) {
...
+    }
+}
+
 function handleBuildSettings (platformConfig, locations, infoPlist) {
     var pkg = platformConfig.getAttribute('ios-CFBundleIdentifier') || platformConfig.packageName();
     var targetDevice = parseTargetDevicePreference(platformConfig.getPreference('target-device', 'ios'));
     var deploymentTarget = platformConfig.getPreference('deployment-target', 'ios');
     var needUpdatedBuildSettingsForLaunchStoryboard = checkIfBuildSettingsNeedUpdatedForLaunchStoryboard(platformConfig, infoPlist);
     var swiftVersion = platformConfig.getPreference('SwiftVersion', 'ios');
+    var displayName = platformConfig.name().replace(/"/g, '');
```
[the remark this answers is not in the document]
Comment (NiklasMerz) [reply bin/templates/scripts/cordova/lib/prepare.js]: I did and quick searcg and I could not find a way to get `projectName` or `targetName`. But these lines use displayName and shortName. Should I change this to short name? 

https://github.com/GEDYSIntraWare/cordova-ios/blob/55a06f97bbc72769afa0d32b97053357ce5dba1d/bin/templates/scripts/cordova/lib/prepare.js#L190

Shortname falls back to the full name if not defined:
https://github.com/apache/cordova-common/blob/7ceeeb4ddf0930343588c77c09561a6b4aa89b4b/src/ConfigParser/ConfigParser.js#L121

And this needs to be encoded with `unorm.nfd()`. 

What do you suggest?

**Comment by NiklasMerz** [comment]
I changed the the original PR that way that only the bundle id gets checked for the target. Other settings should be applied to other targets, too.

For me type extensions this necessary. @msari-ipe-ext-1 Does this work for watch extensions.

**Comment by NiklasMerz** [comment]
Would be great to get this and this https://github.com/apache/cordova-node-xcode/pull/79 fixed for the next release. How can I help to get this ready?

**Comment by brodybits** [comment]
Hi @NiklasMerz I may need some extra time to understand the history of issue #538, previous attempts such as #545, this proposal, and apache/cordova-node-xcode#79, before I am ready to answer your question. I would also like to point to the sustainability issue I raised in apache/cordova#163. Thanks for your understanding and thanks for your recent contributions!

File: bin/templates/scripts/cordova/lib/prepare.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (brodybits) [reply bin/templates/scripts/cordova/lib/prepare.js]: I do not follow the arguments 100%. One thing is that one of the code references seems to point into a fork, please point into the code in our repos whenever possible.

@NiklasMerz can you please give a quick explanation of what this variable means?

**Comment by brodybits** [comment]
I added [MAJOR] to the title and set the milestone to the next major release (6.0.0), due to the risk of this change breaking something else.

File: bin/templates/scripts/cordova/lib/prepare.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (NiklasMerz) [reply bin/templates/scripts/cordova/lib/prepare.js]: Oof this was a while ago and I am having a hard time to understand my comments, too. Should be a sign that I am not clear enough.

So the line in our fork is the same in the original repo. I probably got confused with the repos:
https://github.com/apache/cordova-ios/blob/c1a6640758d1e95f7f7e0dbca980df8780a2adf4/bin/templates/scripts/cordova/lib/prepare.js#L191

I am just realising @dpogue probably suggest to change the name of variable and I thought about getting another variable. That's why it's so confusing. Thank you helping sort that out. I think that makes sense and will clear that up once I am back at my workstation.

**Comment by NiklasMerz** [comment]
Thanks @brodybits for your review. Good points.

I am aware of the current maintainer situation. I will do my best to clear this code up and explain the open questions as soon as I find time at work with my dev environment on hand.

Tests are a good idea since this should be a fix for a regression as described in #583. To put that confusing discussion into short words: Cordova previously did not touch additional targets in Xcode projects (like Watch extension, call directroy extension ..). With Cordova 5 that changed and thos additinal targets get the name of the host app for example. This is undesirable for certain complex apps with additional targets.

This PR tries to bring back the old behavior basically. I think the whole replacing placeholder stuff in Xcode projects changed in Cordova 5. That's why https://github.com/apache/cordova-node-xcode/pull/79 is needed to fix the `PRODUCT_BUNDLE_IDENTIFIER` placeholder while adding targets with hooks. 

As said before I am going to work on the code and I am open to discussions. We are using to extensions to our Cordova app and this is really important. A new release with this fix is not urgent because using forks works fine for a while but I don't want to maintain these forks forever.

**Comment by NiklasMerz** [comment]
I updated the `displayName` variable to make this clear.

I can agree the `updateBuildPropertyLocal` method is not that easy to understand but it is handling info.plist file wizardry I don't understand either. The method is written by @wolfmanfx in #545 and just put together in this PR.

@wolfmanfx What do you think of this PR? Do you have ideas how to improve this method?

**Comment by brodybits** [comment]
Thanks @NiklasMerz for the recent fix, looks nice. And thanks for your help and understanding of our maintainer situation.

I completely understand that you are basically adapting a bug fix from someone else, with the obvious challenges of taking over code with missing test cases.

I would like to continue the discussion in #538 which I have just pinned in order to get a better understanding of the whole situation of this functionality.

**Comment by NiklasMerz** [comment]
Yes the missing test cases part is crucial. That kind of app is not considered in the tests and I am having a hard time thinking of adding it right now.

But I must admit that using extensions with Cordova might not be a often used features except for plugins like [open with](https://github.com/j3k0/cordova-plugin-openwith) .

**Comment by timbru31** [comment]
I've restarted the timed out Travis test, CI is now green

File: bin/templates/scripts/cordova/lib/prepare.js [hunk 2]
```diff
@@ -275,12 +275,54 @@ function handleOrientationSettings (platformConfig, infoPlist) {
     }
 }
 
+// Make sure only update properties from our target project
+function updateBuildPropertyLocal (proj, targetName, prop, value, build) {
+    try {
+        // Check if we have a valid target - during prepare we do not have it
+        var target = proj.pbxTargetByName(targetName);
```
Comment (erisu) [inline bin/templates/scripts/cordova/lib/prepare.js]:
```suggestion
        const target = proj.pbxTargetByName(targetName);
```

File: bin/templates/scripts/cordova/lib/prepare.js [hunk 3]
```diff
@@ -275,12 +275,54 @@ function handleOrientationSettings (platformConfig, infoPlist) {
...
+        if (target == null || target.buildConfigurationList == null) {
```
Comment (erisu) [inline bin/templates/scripts/cordova/lib/prepare.js]:
```suggestion
        if (!target || !target.buildConfigurationList) {
```
If you want to use `== null`, make sure you use triple equals `=== null`

File: bin/templates/scripts/cordova/lib/prepare.js [hunk 4]
```diff
@@ -275,12 +275,54 @@ function handleOrientationSettings (platformConfig, infoPlist) {
...
+            proj.updateBuildProperty(prop, value, build);
+        } else {
+            var targetProjectBuildReference = target.buildConfigurationList;
+            // Collect the uuid's from the configuration of our target
+            var COMMENT_KEY = /_comment$/;
+            var validConfigs = [];
+            var configList = proj.pbxXCConfigurationList();
```
Comment (erisu) [inline bin/templates/scripts/cordova/lib/prepare.js]: Try to use `const` and `let` here as well.. any anywhere else..

Comment (erisu) [reply bin/templates/scripts/cordova/lib/prepare.js]: Just a nitpick but `targetProjectBuildReference` is used only once and also it is 2 characters less than `target.buildConfigurationList`, you might not actually need to create an extra variable.

File: bin/templates/scripts/cordova/lib/prepare.js [hunk 5]
```diff
@@ -275,12 +275,54 @@ function handleOrientationSettings (platformConfig, infoPlist) {
...
+                        validConfigs.push(buildVariants[i].value);
+                    }
+                    break;
+                }
+            }
+            // Only update target props
+            var configs = proj.pbxXCBuildConfigurationSection();
+            for (configName in configs) {
+                if (!COMMENT_KEY.test(configName)) {
+                    if (validConfigs.indexOf(configName) === -1) {
```
Comment (erisu) [inline bin/templates/scripts/cordova/lib/prepare.js]:
```suggestion
                    if (!validConfigs.includes(configName)) {
```

File: bin/templates/scripts/cordova/lib/prepare.js (hunk 3 above)
Comment (NiklasMerz) [reply bin/templates/scripts/cordova/lib/prepare.js]: I am working to much with typescript. I have forgotton all JS quirks :-)

File: bin/templates/scripts/cordova/lib/prepare.js (hunk 4 above)
Comment (NiklasMerz) [reply bin/templates/scripts/cordova/lib/prepare.js]: Done

**Comment by NiklasMerz** [comment]
This code somehow works for my project, but is not correct. We need to find a better solution possibly involving a change in cordova-node-xcode.

I would love to contribute this but I think it is beyon my capabilities and understandings of Cordova. I tried to work on this a few times but I head always explodes. I appreciate if someone with more experience has the capacity to work on this and I will do my best to support that.

**Pull request closed without merging**